controllers/apply-next/helpers.js

```diff
@@ -1,186 +1,268 @@
+            const errorsForSection = errors.filter(detail => includes(fieldNames, detail.context.key));
+
+            obj[section.slug] = determineStatus(dataForSection, errorsForSection);
+
+            return obj;
+        }, {})
+    };
+}
 
-    // Check whether the entire form is empty
-    const formIsNotEmpty = clonedForm.sections.some(section => section.state !== FORM_STATES.empty);
+/**
+ * @typedef {object} MatchOptions
+ * @property {String} baseUrl
+ * @property {Array} sections
+ * @property {Number} currentSectionIndex
+ * @property {Number} currentStepIndex
+ * @property {Object} formData
+ */
 
-    if (formIsNotEmpty) {
-        // Work out the entire form's state based on its sections' state
-        const formHasInvalidSections = clonedForm.sections.some(section => section.state !== FORM_STATES.complete);
-        clonedForm.state = formHasInvalidSections ? FORM_STATES.incomplete : FORM_STATES.complete;
+/**
+ * Find next matching URL
+ * @param {MatchOptions} options
+ */
+function findNextMatchingUrl({ baseUrl, sections, currentSectionIndex, currentStepIndex, formData }) {
+    const currentSection = sections[currentSectionIndex];
+    const nextSection = sections[currentSectionIndex + 1];
+
+    const targetStepIndex = findIndex(
+        currentSection.steps,
+        step => (step.matchesCondition ? step.matchesCondition(formData) === true : true),
+        currentStepIndex + 1
+    );
+
+    if (targetStepIndex !== -1 && targetStepIndex <= currentSection.steps.length) {
+        return `${baseUrl}/${currentSection.slug}/${targetStepIndex + 1}`;
+    } else if (nextSection) {
+        return `${baseUrl}/${nextSection.slug}`;
     } else {
-        clonedForm.state = FORM_STATES.empty;
+        return `${baseUrl}/summary`;
     }
-    return clonedForm;
 }
 
-function getFieldsForStep(step) {
-    return flatMap(step.fieldsets, fieldset => fieldset.fields);
+/**
+ * Find previous matching URL
+ * @param {MatchOptions} options
+ */
+function findPreviousMatchingUrl({ baseUrl, sections, currentSectionIndex, currentStepIndex, formData }) {
+    const currentSection = sections[currentSectionIndex];
+    const previousSection = sections[currentSectionIndex - 1];
+
+    if (currentStepIndex !== 0) {
+        const targetStepIndex = findLastIndex(
+            currentSection.steps,
+            step => (step.matchesCondition ? step.matchesCondition(formData) === true : true),
+            currentStepIndex - 1
+        );
+        return `${baseUrl}/${currentSection.slug}/${targetStepIndex + 1}`;
+    } else if (previousSection) {
+        return `${baseUrl}/${previousSection.slug}/${previousSection.steps.length}`;
+    } else {
+        return baseUrl;
+    }
 }
 
-function getFieldsForSection(section) {
-    return flatMap(section.steps, getFieldsForStep);
+/**
+ * Find next and previous matching URLs
+ * @param {MatchOptions} options
+ */
+function nextAndPrevious(options) {
```

Comment:
Refactored the "determine next step" logic to be usable to work out both the next and previous url. Added tests for this logic.